gadopt/gplates.py

```diff
@@ -0,0 +1,449 @@
+            ((1e-2 * pyGplatesConnector.velocity_non_dim_factor) / (self.scaling_factor * pyGplatesConnector.yrs2sec))
+
+        # generate a KD-tree of the seeds points that have a numerical value
+        tree = cKDTree(data=self.seeds[seeds_u[:, 0] == seeds_u[:, 0], :], leafsize=16)
+
+        # find the neighboring points
+        dists, idx = tree.query(x=target_coords, k=self.nneighbours)
+
+        # Use np.where to avoid division by zero
+        # Replace 0 distances with np.finfo(float).eps to avoid division
+        # by zero while keeping original zeros intact for later use
+        safe_dists = np.where(dists == 0, np.finfo(float).eps, dists)
+
+        # Then, calculate the weighted average using safe_dists
+        res_u = np.einsum(
+            'i, ij->ij',
+            1/np.sum(1/safe_dists, axis=1),
+            np.einsum('ij, ijk ->ik', 1/safe_dists, seeds_u[seeds_u[:, 0] == seeds_u[:, 0]][idx])
+        )
+
+        # Now handle the case where points are too close to each other:
+        res_u[dists[:, 0] <= 1e-8, :] = seeds_u[seeds_u[:, 0] == seeds_u[:, 0]][idx[dists[:, 0] <= 1e-8, 0]]
+
+        return res_u
+
+    def _calc_velocities(self, velocity_domain_features, topology_features, rotation_model, time, delta_time):
+        """
+        Calculates velocity vectors for domain points at a specific geological time.
+
+        This method calculates the velocities for all points in the velocity domain
+        at the specified geological time, using pyGplates to account for tectonic plate
+        motions. It utilizes a plate partitioner to determine the tectonic plate each
+        point belongs to and computes the velocity based on the rotation model and
+        the change over the specified time interval.
+
+        Args:
+            velocity_domain_features: Domain features representing the points at which velocities are to be calculated.
+            topology_features: Topological features of tectonic plates used in the partitioning.
+            rotation_model: The rotation model defining plate movements over time.
+            time (float): The geological time at which velocities are to be calculated.
+            delta_time (float): The time interval over which velocity is calculated.
+
+        Returns:
+            list of pygplates.Vector3D: Velocity vectors for all domain points.
+
+        Raises:
+            RuntimeWarning: If no plate ID is found for a given point, indicating an issue with the reconstruction model.
+        """
+        # All domain points and associated (magnitude, azimuth, inclination) velocities for the current time.
+        all_domain_points = []
+        all_velocities = []
+
+        # Partition our velocity domain features into our topological plate polygons at the current 'time'.
+        # Note: pygplates can only deal with rounded ages in million years
+        plate_partitioner = pygplates.PlatePartitioner(topology_features, rotation_model, round(time, 0))
```

Comment:
Given you use `round(time, 0)` several times, you could factor that out.

Reply:
Makes sense!